File: OSMiniProject1/Scheduler.cpp
#include "Scheduler.h"



void Scheduler::HPF()
{
	while (!done)
	{
		refreshProcesses();
		// algo
		if (activeProcesses.size() == 1)
		{

		}
		// algo end
		time++;
	}
}

void Scheduler::FCFS()
{
	int start = 0;
	while (numOfProcesses)
	{
		refreshProcesses();
		// algo
	
		for (int i = start; i < activeProcesses.size(); i++)
		{
			float burst = activeProcesses[i].getburstTime();
			activeProcesses[i].setFinishingTime(burst + time);
			start++;
			busy(burst);
			time++; // assume context time = 1
			numOfProcesses--;
		}
		time++;
		// algo end
	}
	// To be deleted 
	cout << endl << "processes algo done" << endl;
	for (int i = 0; i < activeProcesses.size(); i++)
	{
		cout << activeProcesses[i].getArrivalTime() << "\t" <<activeProcesses[i].getburstTime()<<"\t" << activeProcesses[i].getFinishingTime() << endl;
	}
}

void Scheduler::RR()
{
}

void Scheduler::SRTN()
{
}

void Scheduler::busy(int t)
{
	for (int i = 0; i < t; i++)
	{
		time++;
		refreshProcesses();
	}
}

void Scheduler::refreshProcesses()
{
	for (int i = 0; i < processes.size(); i++)
	{
		if (processes[i].getArrivalTime() == time) {
			activeProcesses.push_back(processes[i]);
		}
		else if (processes[i].getArrivalTime() > time)
		{
			break;
		}
	}
}

Scheduler::Scheduler()
{
	done = false;
	time = 0;
}

void Scheduler::addToScheduler(string s)
{
	ifstream data;
	data.open(s, ios::in);
	if (data.is_open())
	{
		data >> numOfProcesses;
		for (int i = 0; i < numOfProcesses; i++)
		{
			float arrivalTime;
			float	burstTime;
			int priority;
			int id;
			data >> id >> arrivalTime >> burstTime >> priority;
			processes.push_back(Process(id , arrivalTime, burstTime, priority));
		}
	}

}

void Scheduler::sort()
{
	
	for (int i = 1; i < processes.size(); i++)
	{
		float key = processes[i].getArrivalTime();
		Process k = processes[i];
		int j = i - 1;
		while (j >= 0 && ((processes[j].getArrivalTime() > key)||(processes[j].getArrivalTime() == key && k.getPriority()> processes[j].getPriority()))) {
			processes[j + 1] = processes[j];
			j--;
		}
		processes[j + 1] = k;
	}
	// To be deleted 
	cout <<endl<< "processes sorted" << endl;
	for (int i = 0; i < processes.size(); i++)
	{
		cout << processes[i].getArrivalTime()<<"\t" << processes[i].getPriority()<< endl;
	}
}

void Scheduler::start(int algo)
{
	switch (algo)
	{
	case 1:
		HPF();
		break;
	case 2:
		FCFS();
		break;
	case 3:
		RR();
		break;
	case 4:
		SRTN();
		break;
	default:
		break;
	}
}


Scheduler::~Scheduler()
{
}

File: python/Scheduler.py
# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""
from matplotlib import pyplot as plt
from matplotlib import style

from Process import *
class Scheduler():
   
    def __init__(self):
        self.NumOfProcess = 0
        self.Processes = list()
        self.ActiveProcesses = list()
        self.PrintProcesses = list()
        self.PrintInfo = list()
        self.Done = False
        self.Time = 0
        self.context = 0
        self.quantum = 4
        
    def Draw(self):
        style.use('ggplot')

        for i in self.PrintProcesses:
            plt.bar(i.GetArrival(), i.GetID(), width=i.GetBurstTime(), align='edge')

        plt.show()
#####################################################################################################################################
    def printData(self):
        with open("Data.txt" , 'w+') as file:
            file.write( "number of processes" + str(len(self.PrintInfo)))
            file.write('\n' + "Id " + "wT " + "TAT " + "weightedTAT" )
            AVGTAT = 0
            AVGWTAT= 0
            for i in range(len(self.PrintInfo)):
                file.write('\n' +str(self.PrintInfo[i].GetID()) + " " + str(self.PrintInfo[i].GetWaitingTime()) + " " + str(self.PrintInfo[i].GetTurnAroundTime()) + " " +  str(self.PrintInfo[i].GetWeightedTat()))
                AVGTAT+=self.PrintInfo[i].GetTurnAroundTime()
                AVGWTAT+=self.PrintInfo[i].GetWeightedTat()
            AVGTAT /=len(self.PrintInfo)
            AVGWTAT /=len(self.PrintInfo)
            file.write('\n' +"Average TAT: " + " " + str(AVGTAT) + " " )
            file.write('\n' +"Average weighted TAT: " + " " + str(AVGWTAT) + " " )    

##############################################################################################
    def start(self , argument , cont , quant):
         if cont != "":
            self.context  = int(cont) 
         if quant != "":
            self.quantum = int(quant)
         switcher = {
         1: "HPF",
         2: "FCFS",
         3: "RR",
         4: "SRTN",
         }
         x = switcher.get(argument, "Invalid algo")
         
         if(x=="HPF"):
             self.HPF()
             print('HPF is choosen')
         elif(x=="FCFS"):
            self.FCFS()  ### function of FCFS
            print('FCFS is choosen')
         elif(x=="RR"):
             self.RR()
             x=3 ### function of RR
             print('RR is choosen')
         elif(x=="SRTN"):
            self.SRTN()
            print('SRTNs is choosen')
            
##############################################################################################
    def FCFS(self):
        while self.NumOfProcess:
            self.RefreshProcesses()
            while len(self.ActiveProcesses):
                brust = self.ActiveProcesses[0].GetBurstTime();
                self.ActiveProcesses[0].SetFinishTime(brust + self.Time)
                self.PrintInfo.append(self.ActiveProcesses[0])
                self.Busy(brust)
                # for easy the DRAW 
                self.PrintProcesses.append(Process(self.ActiveProcesses[0].GetID(),self.Time-brust,brust,self.Time))
                #####################
                self.ActiveProcesses.pop(0)
                self.NumOfProcess -= 1
                self.Busy(self.context)
            self.Time+=1
        self.printData()
        self.Draw()

    def RR(self):
         while self.NumOfProcess:
             self.RefreshProcesses()
             while len(self.ActiveProcesses):
                remaining = self.ActiveProcesses[0].GetRemainingTime();
                if(remaining >= self.quantum):
                    id = self.ActiveProcesses[0].GetID()
                    self.ActiveProcesses[0].SetRemainingTime(self.ActiveProcesses[0].GetRemainingTime()-self.quantum)
                    self.Busy(self.quantum)
                    self.ActiveProcesses.append(self.ActiveProcesses[0])
                    self.ActiveProcesses.pop(0)
                    self.PrintProcesses.append(Process(id,self.Time-self.quantum,self.quantum,1))
                    if id != self.ActiveProcesses[0].GetID():
                        self.Busy(self.context)
                else:
                    self.ActiveProcesses[0].SetFinishTime(remaining + self.Time)
                    self.Busy(remaining)
                    self.PrintInfo.append(self.ActiveProcesses[0])
                

                    self.PrintProcesses.append(Process(self.ActiveProcesses[0].GetID(),self.Time-remaining,remaining,self.Time))
                    self.NumOfProcess -= 1
                    self.ActiveProcesses.pop(0)
                    self.Busy(self.context)
             self.Time+=1
         self.printData()
         self.Draw()
##############################################################################################################################
    def AddToScheduler(self , InputFile):
         with open(InputFile) as file:
            self.NumOfProcess = int(file.readline())
            content = file.readlines()
            for pro in content:
                ID , arrival , burst , priority = map(int, pro.split(" "))
                self.Processes.insert( ID-1 , Process( ID ,arrival , burst, priority ))
    
    def SortTime(self):
         self.Processes.sort( key = lambda x: (x.GetArrival() , x.GetID()))
    
    def SortPriority(self):
       self.ActiveProcesses.sort( key = lambda x: (x.GetPriority() , -x.GetID() ) , reverse=True)
    def SortRemainingTime(self):
       self.ActiveProcesses.sort( key = lambda x: (x.GetRemainingTime() , x.GetID() ) )
       
    def RefreshProcesses(self):
        for pro in self.Processes:
            if(pro.GetArrival() == self.Time):
                self.ActiveProcesses.append(pro)
            elif (pro.GetArrival() > self.Time):
                break
    def Busy(self , clock):
        for i in range(clock):
            self.Time+=1
            self.RefreshProcesses()
            
    def HPF(self):
       
        while self.NumOfProcess:
            self.RefreshProcesses()
            self.SortPriority()
            while len(self.ActiveProcesses):
                brust = self.ActiveProcesses[0].GetBurstTime();
                self.ActiveProcesses[0].SetFinishTime(brust + self.Time)
                self.PrintInfo.append(self.ActiveProcesses[0])
                self.Busy(brust)
                self.PrintProcesses.append(Process(self.ActiveProcesses[0].GetID(),self.Time-brust,brust,self.Time))
                self.ActiveProcesses.pop(0)
                self.NumOfProcess -= 1
                
                self.SortPriority()
                self.Busy(self.context)
              
            self.Time+=1
        self.printData()
        self.Draw()    
                
            # kharaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa############################
    def SRTN(self):
       
        while self.NumOfProcess:
            self.RefreshProcesses()
            self.SortRemainingTime()
            while len(self.ActiveProcesses):
                id = self.ActiveProcesses[0].GetID()
                p = self.ActiveProcesses[0]
                end = True
                counter =0 
                while self.ActiveProcesses[0].GetRemainingTime():
                    self.Time+=1
                    counter+=1
                    p.SetRemainingTime(p.GetRemainingTime() - 1)
                    self.RefreshProcesses()
                    self.SortRemainingTime()
                    if id!=self.ActiveProcesses[0].GetID():
                       self.PrintProcesses.append(Process(p.GetID(),self.Time-counter,counter,self.Time)) 
                       self.Busy(self.context)
                       end= False
                       break
                if(end):
                    self.ActiveProcesses[0].SetFinishTime(self.Time)
                    self.PrintInfo.append(self.ActiveProcesses[0])
                    self.PrintProcesses.append(Process(self.ActiveProcesses[0].GetID(),self.Time-counter,counter,self.Time))
                    self.NumOfProcess -= 1
                    self.ActiveProcesses.pop(0)
                    self.Busy(self.context)
            self.Time+=1
        self.printData()
        self.Draw()  

        
            

    
        
    
       
#g = Scheduler()     
#g.AddToScheduler("processes.txt")
#g.SortTime()
#g.RefreshProcesses()

#print(g.NumOfProcess)
#for x in g.ActiveProcesses:
#    print(x.GetID()  , x.GetArrival() , x.GetBurstTime(), x.GetPriority() )
  
        
